### PR TITLE
Prototype for 1658

### PR DESCRIPTION
**Pull request by sbrossie**
Depends on:
* https://github.com/killbill/killbill-api/pull/90: API PR
* https://github.com/killbill/killbill-plugin-api/pull/22: Plugin api PR

Main commits to review:

* https://github.com/killbill/killbill/pull/1676/commits/fa71142082a0d7e282cdb4ad9bc1b2a7fa814247: Implementation for `InvoicePluginDispatcher#splitInvoices`
*https://github.com/killbill/killbill/pull/1676/commits/162f871e15409878aa369963b5c70634d112db87:  `Invoice#grpId`
* https://github.com/killbill/killbill/pull/1676/commits/7e7574827cb9d9c382979343d8122aa03f07f040: JAXRS sendpoints
* https://github.com/killbill/killbill/pull/1676/commits/de478d157e5139c2b866ce62d6355569d748bf21 and https://github.com/killbill/killbill/pull/1676/commits/349f9c99c62748ea5f467ed81c46b07555ae4b4d: Restriction form ability to split invoices.

There are a few `TODO_1658` to look at prior merge probably.

TODO: Need to regenerate java client and write a small `killbill` profiles test.

### CODE REVIEW

**Comment by pierre** [comment]
🛑 ⚠️ **DO NOT MERGE** ⚠️ 🛑 

Keeping this PR open while there are still comments to be addressed.

It will be automatically closed once https://github.com/killbill/killbill/pull/1683 is merged.

File: invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java [hunk 1]
```diff
@@ -173,6 +182,52 @@ private void onCompletionCall(final boolean isSuccess,
...
+                                                                          public SimpleEntry<UUID, InvoiceItem> apply(final InvoiceItem invoiceItem) {
+                                                                              return new SimpleEntry<>(invoiceItem.getId(), invoiceItem);
+                                                                          }
+                                                                      }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+                final List<InvoiceGroup> groups = grpResult.getInvoiceGroups();
+                for (final InvoiceGroup grp : groups) {
+                    final DefaultInvoice grpInvoice = (DefaultInvoice) originalInvoice.clone();
+                    final UUID newInvoiceId = UUIDs.randomUUID();
+                    grpInvoice.setId(newInvoiceId);
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java:213]: Nit: maybe we could introduce a `cloneForGrouping` which hides that cleanup.

**Review by pierre — COMMENTED** [review]
👍 Just a few thoughts below.

File: invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java [hunk 2]
```diff
@@ -306,6 +306,29 @@ public InvoiceModelDao inTransaction(final EntitySqlDaoWrapperFactory entitySqlD
         });
     }
 
+    @Override
+    public List<InvoiceModelDao> getInvoicesByGroup(final UUID groupId, final InternalTenantContext context) {
+        // TODO_1658 Do we want a special query along with a new index on grpId or is the cardinality small enough that we
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java:311]: As discussed, I will create the missing index on `invoice#grp_id` so we can move to a separate query when/if the time comes.

File: invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java [hunk 3]
```diff
@@ -395,14 +402,19 @@ public List<InvoiceItemModelDao> inTransaction(final EntitySqlDaoWrapperFactory
...
                     final boolean isNotShellInvoice = invoiceIdsReferencedFromItems.remove(invoiceModelDao.getId());
 
                     final InvoiceModelDao invoiceOnDisk = existingInvoiceMetadata.getExistingInvoice(invoiceModelDao.getId(), context);
                     if (isNotShellInvoice) {
                         // Create the invoice if this is not a shell invoice and it does not already exist
                         if (invoiceOnDisk == null) {
+                            if (grpId == null) {
+                                grpId = invoiceModelDao.getId();
+                            }
+                            invoiceModelDao.setGrpId(grpId);
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java:417]: The default value in the CTOR is for when there is no group, i.e a group of 1: In which case, the `invoice#grp_id` is the `invoice#id`. Here, we actually set the id of the group for when there is a real group by defaulting to the `invoice#id` associated with the first invoice in the group.

File: jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/InvoiceResource.java [hunk 4]
```diff
@@ -191,6 +192,35 @@ public Response getInvoice(@PathParam("invoiceId") final UUID invoiceId,
...
+        final List<InvoiceJson> result = new ArrayList<>();
+        final Iterator<Invoice> it = invoices.iterator();
+        while (it.hasNext()) {
+            final Invoice invoice  = it.next();
+            final List<InvoiceItem> childInvoiceItems = withChildrenItems ? invoiceApi.getInvoiceItemsByParentInvoice(invoice.getId(), tenantContext) : null;
+            final AccountAuditLogs accountAuditLogs = auditUserApi.getAccountAuditLogs(invoice.getAccountId(), auditMode.getLevel(), tenantContext);
+            result.add(new InvoiceJson(invoice, childInvoiceItems, accountAuditLogs));
+        }
+        if (result.isEmpty()) {
+            return Response.status(Status.NOT_FOUND).build();
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/InvoiceResource.java:218]: We are consistent with the existing  `createFutureInvoice` which returns `NOT_FOUND` if nothing was generated. Are you referring to other endpoints that would return a list of things - which one?

File: invoice/src/main/java/org/killbill/billing/invoice/api/user/DefaultInvoiceUserApi.java [hunk 5]
```diff
@@ -272,20 +279,32 @@ public List<Invoice> getUnpaidInvoicesByAccountId(final UUID accountId, final Lo
...
         final InternalCallContext internalContext = internalCallContextFactory.createInternalCallContext(accountId, context);
 
-        final Invoice result = dispatcher.processAccount(true, accountId, targetDate, dryRunArguments, false, internalContext);
-        if (result == null) {
+        final List<Invoice> result = dispatcher.processAccount(true, accountId, targetDate, dryRunArguments, false, false, internalContext);
+        if (result.isEmpty()) {
             throw new InvoiceApiException(ErrorCode.INVOICE_NOTHING_TO_DO, accountId, targetDate != null ? targetDate : "null");
         } else {
-            return result;
+            Preconditions.checkState(result.size() == 1);
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/api/user/DefaultInvoiceUserApi.java:286]: See https://github.com/killbill/killbill/pull/1687

File: invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/dao/DefaultInvoiceDao.java:311]: See https://github.com/killbill/killbill/pull/1687

File: invoice/src/main/java/org/killbill/billing/invoice/model/DefaultInvoice.java [hunk 6]
```diff
@@ -320,6 +323,11 @@ public UUID getParentInvoiceId() {
         return parentInvoice != null ? parentInvoice.getId() : null;
     }
 
+    @Override
+    public UUID getGroupId() {
+        return grpId;
+    }
+
     @Override
     public String toString() {
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/model/DefaultInvoice.java:332]: See https://github.com/killbill/killbill/pull/1687

File: invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java (hunk 1 above)
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java:213]: I went back and forth and decided to simply reuse a [CTOR that already exists](https://github.com/killbill/killbill/pull/1687/commits/c133922b00dbe10ee89cd5b6e6897299d0719b7c#diff-a6f829e33b0216520626dd03b04a2a7ac9c9018a7e39bbafb09e4e9c36de0c60R209).

**Pull request merged**